### PR TITLE
deps: Dump/interpolate repository_locations -> json

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:

There are a few tools that require the repository_locations data, currently we are bringing that data into the python by putting the implementation (as opposed to the abstract library) in the envoy repo, this will allow us to instead pass the repo locations data as a file/cli flag, and move the implementation into the upstream tooling

This PR also avoids using the old `SourceFileLoader` + bzl files method of loading the data, instead loading directly from the `write_json` rules for the repo files and interpolating

There is a more generic implementation of some of what is required here in #17998

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
cc @htuch @kfaseela

File: tools/dependency/dump_repository_locations.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    for repo_file in repo_location_files:
+        repo_locations = json.loads(pathlib.Path(repo_file).read_text())
+        for location in repo_locations.values():
+            if "version" in location:
+                if "strip_prefix" in location:
+                    location["strip_prefix"] = _format_version(
+                        location["strip_prefix"], location["version"])
+                location["urls"] = [
+                    _format_version(url, location["version"]) for url in location["urls"]
+                ]
```
Comment (htuch) [inline tools/dependency/dump_repository_locations.py:25]: Why can't this be done more intelligently during the generation of the input JSON files? 

I used to have some Python code that was usable in both Bazel and Python in the `.bzl` and that centralized this logic.

Comment (phlax) [reply tools/dependency/dump_repository_locations.py:25]: so, bazel can provide `json` files, but we need a couple of extra steps here - interpolating the urls/versions, and aggregating from several `repository_locations.bzl` files

as mentioned in the ticket comment i have a more generic solution for this based on `data` providers, and a mutation pipeline - it was designed to work in several places, here being one of them

the old code that used `SourceFileLoader` was problematic for a few reasons - eg you cant put `load` files in the `.bzl` - ultimately i dont think its a good idea to parse starlark using a python parser

also, it wasnt being done in one place, i have been gradually unpicking the places that used this pattern, firstly the docs, and now here - removal is almost complete

Comment (phlax) [reply tools/dependency/dump_repository_locations.py:25]: the way the `SourceFileLoader` code is setup it causes another problem also, the paths are different when a job is `run` or `build` 

shifting this to a `py_binary` with `data` and `location` `args` kinda resolves that

Comment (phlax) [reply tools/dependency/dump_repository_locations.py]: ok, so i think i understand better your suggestion

i have made the `json_data` write rules for `repository_locations` do the interpolation in bazel using the existing bazel util

this removes the need to do the mangling in python, but also the need to load any starlark as python

Comment (phlax) [reply tools/dependency/dump_repository_locations.py]: to make it work i also had to add a `merge_dicts` util as there is no one-liner with starlark python to merge dicts in place

Comment (htuch) [reply tools/dependency/dump_repository_locations.py]: wh00t, thanks!

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**